streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```
@@ -0,0 +1,178 @@
+ */
+package org.apache.kafka.streams.processor.internals;
+
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.errors.TimeoutException;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.streams.errors.StreamsException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class StoreChangelogReader implements ChangelogReader {
+    private static final Logger log = LoggerFactory.getLogger(StoreChangelogReader.class);
+
+    private final Consumer<byte[], byte[]> consumer;
+    private final Time time;
+    private final long partitionValidationTimeoutMs;
+    private final Map<String, List<PartitionInfo>> partitionInfo = new HashMap<>();
+    private final Map<TopicPartition, StateRestorer> stateRestorers = new HashMap<>();
+
+
+    public StoreChangelogReader(final Consumer<byte[], byte[]> consumer, final Time time, final long partitionValidationTimeoutMs) {
+        this.consumer = consumer;
+        this.time = time;
+        this.partitionValidationTimeoutMs = partitionValidationTimeoutMs;
+    }
+
+
+    @Override
+    public void validatePartitionExists(final TopicPartition topicPartition, final String storeName) {
+        final long start = time.milliseconds();
+        if (partitionInfo.isEmpty()) {
+            try {
+                partitionInfo.putAll(consumer.listTopics());
+            } catch (final TimeoutException e) {
+                throw new StreamsException(String.format("Could not fetch partition info for topic: %s before expiration of the configured request timeout",
```

Should we really throw an exception here since it is only for optimization? Otherwise if this step times out we will never continue to the following step.

---

sure will just log it